packages/core/admin/admin/src/utils/fetchClient.js

```diff
@@ -0,0 +1,41 @@
+import axios from 'axios';
+import { auth } from '@strapi/helper-plugin';
+
+const instance = axios.create({
+  baseURL: process.env.STRAPI_ADMIN_BACKEND_URL,
+});
+
+instance.interceptors.request.use(
+  async (config) => {
+    config.headers = {
+      Authorization: `Bearer ${auth.getToken()}`,
+      Accept: 'application/json',
+      'Content-Type': 'application/json',
+    };
+
+    return config;
+  },
+  (error) => {
+    Promise.reject(error);
+  }
+);
+
+instance.interceptors.response.use(
+  (response) => response,
+  (error) => {
+    // whatever you want to do with the error
+    if (error?.response?.status === 401) {
+      auth.clearAppStorage();
+      window.location.reload();
+    }
+
+    throw error;
+  }
+);
+
+export default {
```

Comment:
Can we make this a named export and export this utility as `getFetchClient`? I'd also prefer to rename the file.

I'm wondering about something: would we export different instances of axios if we'd call this function multiple times? Should we use a singleton pattern instead to make sure `instance` and the interceptors are only generated once?


```js

export function getFetchClient() {
    if (!instance) {
      instance = axios.create();
      instance.interceptors.request.use();
    }
    
    return instance;
}
```

Also: can you please add tests for that utility? :) I'd personally test:

- we only exports HTTP verbs
- in case of a 401 response reload + clearing storage works
- headers are set properly

Reply:
A nice approach (could be usefull for testing) is to export a default fetchClient & the getFetchClient utility as a proprty.

Reply:
```js
import axios from 'axios';
import { auth } from '@strapi/helper-plugin';

let instance;

function getFetchClient() {
  if (instance) {
    return instance;
  }

  instance = axios.create({
    baseURL: process.env.STRAPI_ADMIN_BACKEND_URL,
  });

  instance.interceptors.request.use(
    async (config) => {
      config.headers = {
        Authorization: `Bearer ${auth.getToken()}`,
        Accept: 'application/json',
        'Content-Type': 'application/json',
      };

      return config;
    },
    (error) => {
      Promise.reject(error);
    }
  );

  instance.interceptors.response.use(
    (response) => response,
    (error) => {
      // whatever you want to do with the error
      if (error?.response?.status === 401) {
        auth.clearAppStorage();
        window.location.reload();
      }

      throw error;
    }
  );

  return instance;
}
```

export { getFetchClient }

Reply:
I would suggest this approach to benefit from a custom create for testing and mocking & the native singleton pattern that js modules provide.

```js
export const createFetchClient = () => { /* ... */ };

export default createFetchClient();
```

Reply:
ok, understood thanks @gu-stav

Reply:
I might be wrong, but I don't see the need of re-implementing a singleton pattern here because ES6 modules will create a singleton instance by default, which [will be used in the whole code when imported](https://k94n.com/es6-modules-single-instance-pattern). 

Actually, we made the experiment, we created a hook that returned the Axios instance and instantiate it in one of our components, then instantiated the useFetchClient hook in the same component, and check if the instance exported by our custom hook and the one exported by our useFetchClient hook was the same, and it was. 

The only concern I see with this code now is that if we call the `getFetchClient`, yes, we would be creating another Axios instance, so following the explanation above, in my opinion, we could create the Axios instance outside the getFetchClient function and use this function just like a getter. However, I don't see a bad option to export just the methods and not the instance itself, because if we do so, it might be changed and it would be a change in the global scope of the application because of the singleton pattern (please correct me if I'm wrong somewhere 😅 ). 

So, instead of returning the instance in the getFetchClient I would return the methods to make the request:

```javascript
const instance = axios.create({
    baseURL: process.env.STRAPI_ADMIN_BACKEND_URL,
  });

  instance.interceptors.request.use(
    async (config) => {
      config.headers = {
        Authorization: `Bearer ${auth.getToken()}`,
        Accept: 'application/json',
        'Content-Type': 'application/json',
      };

      return config;
    },
    (error) => {
      Promise.reject(error);
    }
  );

  instance.interceptors.response.use(
    (response) => response,
    (error) => {
      // whatever you want to do with the error
      if (error?.response?.status === 401) {
        auth.clearAppStorage();
        window.location.reload();
      }

      throw error;
    }
  );

const getFetchClient = () => ({
  get: instance.get,
  put: instance.put,
  post: instance.post,
  delete: instance.delete,
  getFetchClient,
});

export default getFetchClient;
```

Reply:
@christiancp100 when you dereference the methods from an object you lost the object context (a.k.a `this`) 

I'm not sure how axios function will respond to this. in general it is much better to use a Facade/Abstract pattern to avoid breaking axios internals that may require this to be bound correctly or you will need to at least do 

```js
get: instance.get.bind(instance)
...
```

Reply:
Or if the goal is completely to decouple our implementation from Axios we could create a wrapper such as:
```javascript
const getFetchClient = () => ({
  get: (url, config) => instance.get(url, config),
  put: (url, data, config) => instance.put(url, data, config),
  ...
```

Reply:
Yes that's my suggestion (1st example)

Reply:
True 😅

Reply:
I have implemented yours suggestions